internal/controller/postgrescluster/snapshots.go

```diff
@@ -0,0 +1,280 @@
+		if isPrimary, known := instance.IsPrimary(); isPrimary && known {
+			primaryInstance = instance
+		}
+	}
+	// Return error if primary instance not found
+	if primaryInstance.Name == "" {
+		return nil, errors.New("Could not find primary instance. Cannot create volume snapshot.")
+	}
+
+	// Find pvc associated with primary instance
+	primaryPvc := corev1.PersistentVolumeClaim{}
+	for _, pvc := range clusterVolumes {
+		pvcInstance := pvc.GetLabels()[naming.LabelInstance]
+		pvcRole := pvc.GetLabels()[naming.LabelRole]
+		if pvcRole == naming.RolePostgresData && pvcInstance == primaryInstance.Name {
+			primaryPvc = pvc
+		}
+	}
+	// Return error if primary pvc not found
+	if primaryPvc.Name == "" {
+		return nil, errors.New("Could not find primary's pgdata pvc. Cannot create volume snapshot.")
+	}
+
+	// generate VolumeSnapshot
+	snapshot, err := r.generateVolumeSnapshot(postgrescluster, primaryPvc,
+		postgrescluster.Spec.Backups.Snapshots.VolumeSnapshotClassName)
+	if err == nil {
+		// Add annotation for associated backup job's UID
+		if snapshot.Annotations == nil {
+			snapshot.Annotations = map[string]string{}
+		}
+		snapshot.Annotations[naming.PGBackRestBackupJobId] = string(backupJob.UID)
+	}
+
+	return snapshot, err
+}
+
+// generateVolumeSnapshot generates a VolumeSnapshot that will use the supplied
+// PersistentVolumeClaim and VolumeSnapshotClassName and will set the provided
+// PostgresCluster as the owner.
+func (r *Reconciler) generateVolumeSnapshot(postgrescluster *v1beta1.PostgresCluster,
+	pvc corev1.PersistentVolumeClaim,
+	volumeSnapshotClassName string) (*volumesnapshotv1.VolumeSnapshot, error) {
+
+	snapshot := &volumesnapshotv1.VolumeSnapshot{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: volumesnapshotv1.SchemeGroupVersion.String(),
+			Kind:       "VolumeSnapshot",
+		},
+		ObjectMeta: naming.ClusterVolumeSnapshot(postgrescluster),
+	}
+	snapshot.Spec.Source.PersistentVolumeClaimName = &pvc.Name
+	snapshot.Spec.VolumeSnapshotClassName = &volumeSnapshotClassName
+
+	snapshot.Annotations = postgrescluster.Spec.Metadata.GetAnnotationsOrNil()
+	snapshot.Labels = naming.Merge(postgrescluster.Spec.Metadata.GetLabelsOrNil(),
+		map[string]string{
+			naming.LabelCluster: postgrescluster.Name,
+		})
+
+	err := errors.WithStack(r.setControllerReference(postgrescluster, snapshot))
+
+	return snapshot, err
+}
+
+// getLatestCompleteBackupJob takes a JobList and returns a pointer to the
+// most recently completed backup job. If no completed backup job exists
+// then it returns nil.
+func getLatestCompleteBackupJob(jobs *batchv1.JobList) *batchv1.Job {
+
+	zeroTime := metav1.NewTime(time.Time{})
+	latestCompleteBackupJob := batchv1.Job{
+		Status: batchv1.JobStatus{
+			Succeeded:      1,
+			CompletionTime: &zeroTime,
+		},
+	}
+	for _, job := range jobs.Items {
+		if job.Status.Succeeded > 0 &&
+			latestCompleteBackupJob.Status.CompletionTime.Before(job.Status.CompletionTime) {
```

Comment:
🌱 This almost looks like an opportunity for something in `slices`, [slices.MinFunc](https://pkg.go.dev/slices#MinFunc), maybe? 🤔 I'm not certain the following is any more legible...

untested:
```go
if len(jobs.Items) == 0 {
  return nil
}
job := slices.MinFunc(jobs.Items, func(a, b batchv1.Job) int {
  if a.Succeeded < 1 {
    return 1
  }
  if b.Succeeded < 1 {
    return -1
  }
  return a.Status.CompletionTime.Compare(b.Status.CompletionTime)
})
if job.Succeeded < 1 {
  return nil
}
return &job
```